From 4be28a7887cfdc9821b7822decb9d033533cd563 Mon Sep 17 00:00:00 2001
From: Manish Dharanenthiran <quic_mdharane@quicinc.com>
Date: Wed, 14 Dec 2022 17:08:53 +0530
Subject: [PATCH] ath12k: Update reg_tpc_info from bss_conf during channel
 switch

During initial vdev start, pwr_env will be updated
during assing vif chanctx, but during channel switch
announcement, pwr_env will NOT be updated from
bss_info in STATION mode. Hence, during channel
switch for getting max power, tpc_info will have
previous bandwidth power levels which leads to
out-of-bound array access of sband chanctx.

Fix this by updating pwr_env during vdev
restart sequence and update the power level based
on the new channel context.

Signed-off-by: Manish Dharanenthiran <quic_mdharane@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 15ac559..39d6ac1 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -250,6 +250,9 @@ static const u32 ath12k_smps_map[] = {
 static int ath12k_mac_vdev_create(struct ath12k *ar, struct ath12k_link_vif *arvif);
 static int ath12k_start_vdev_delay(struct ath12k *ar,
 				   struct ieee80211_vif *vif);
+static void ath12k_mac_parse_tx_pwr_env(struct ath12k *ar,
+                                        struct ath12k_link_vif *arvif,
+                                        struct ieee80211_chanctx_conf *ctx);
 static void ath12k_update_bcn_template_work(struct work_struct *work);
 static void ath12k_update_obss_color_notify_work(struct work_struct *work);
 
@@ -9650,6 +9653,10 @@ ath12k_mac_vdev_config_after_start(struct ath12k_link_vif *arvif,
             chandef->chan->band == NL80211_BAND_6GHZ &&
             (ahvif->vdev_type == WMI_VDEV_TYPE_STA || ahvif->vdev_type == WMI_VDEV_TYPE_AP) &&
             test_bit(WMI_TLV_SERVICE_EXT_TPC_REG_SUPPORT, ar->ab->wmi_ab.svc_map)) {
+
+		if (ahvif->vdev_type == WMI_VDEV_TYPE_STA)
+			ath12k_mac_parse_tx_pwr_env(ar, arvif, &arvif->chanctx);
+
                 ath12k_mac_fill_reg_tpc_info(ar, arvif, &arvif->chanctx);
                 ath12k_wmi_send_vdev_set_tpc_power(ar, arvif->vdev_id,
                                                    &arvif->reg_tpc_info);
-- 
2.17.1

